### ISSUE TITLE
VMAgent missing targets in service discovery

### ISSUE THREAD

**Issue by thunderbird86** [VictoriaMetrics/VictoriaMetrics#4154]
### Describe the bug

VMAgent running on the same instance as Prometheus with the same SerivceAccount, and share scrape_config configmap. But not all targets discovered in VMAgent.

Here is sts for VMAgent:

```
apiVersion: apps/v1
kind: StatefulSet
metadata:
  name: victoria-metrics-agent
spec:
  replicas: 1
  serviceName: prometheus
  template:
     spec:
       - name: victoria-metrics-agent
          image: "victoriametrics/vmagent:v1.87.4"
          workingDir: /
          args:
            - -remoteWrite.tmpDataPath=/tmpData
            - -promscrape.configCheckInterval=30s
            - -envflag.enable=true
            - -envflag.prefix=VM_
            - -loggerFormat=json
```
and part of scrape_config

```
- job_name: foobar-internal-api
  scheme: http
  kubernetes_sd_configs:
    - role: endpoints
  relabel_configs:
  - source_labels: [__meta_kubernetes_namespace]
    action: keep
    regex: '^my_ns$'
  - source_labels: [__meta_kubernetes_endpoints_name]
    action: keep
    regex: '^foobar-internal-api$'
  - source_labels: [__meta_kubernetes_pod_container_port_number]
    action: keep
    regex: "5050"

- job_name: foobar-external-api
  scheme: http
  kubernetes_sd_configs:
    - role: endpoints
  relabel_configs:
  - source_labels: [__meta_kubernetes_namespace]
    action: keep
    regex: '^my_ns$'
  - source_labels: [__meta_kubernetes_endpoints_name]
    action: keep
    regex: '^foobar-external-api$'
  - source_labels: [__meta_kubernetes_pod_container_port_number]
    action: keep
    regex: "5050"
```

Config is shared with prometheus

And here is result in service-discovery page in Prometheus 

```
[foobar-external-api (2 / 846 active targets)]
[foobar-internal-api (1 / 846 active targets)]
```
And in the VMAgent

```
foobar-external-api (0/30 active)
foobar-internal-api (0/19 active)
```

Not all targets found, moreover the is different numbers of discovered targets, even in VMAgent between two similar jobs, and significantly less than in Prometheus. 

Also I didn't found any records about `my_ns` namespace, or `foobar-external/internal-api` pod in [api/v1/targets].

Also I've run kubectl pod `bitnami/kubectl` with serviceName: prometheus, and test the following:
```
$ kubectl -n my_ns get endpoints
NAME                   ENDPOINTS                                                         AGE
foobar-external-api   10.15.181.52:32500,100.99.84.62:32500   127d
foobar-internal-api   10.15.181.53:8095                                          127d

$ kubectl  get endpoints -A | grep -v -i NAME | wc -l
169
```
Which matches with log record in VMAgent 

```
reloaded 169 objects from "https://100.64.0.1:443/api/v1/endpoints" in 0.017s; updated=0, removed=0, added=169, resourceVersion="986734413"  
```

I've more 50 scrape_rules, and now only two has mismatching between VM and Prometheus. About discovered targes in VMAgent, almost each job has its own number of discovered targets, despite that in prometheus they has the same about discovered targets between similar jobs 

### To Reproduce

Not sure how to reproduce

### Version

```
# ./vmagent-prod --version
vmagent-20230326-001857-tags-v1.87.4-0-g87769b36d
```

### Used command-line flags

  - -remoteWrite.tmpDataPath=/tmpData
  - -promscrape.configCheckInterval=30s
  - -envflag.enable=true
  - -envflag.prefix=VM_
  - -loggerFormat=json

**Comment by thunderbird86** [comment]
@zekker6, no.

Here is from prometheus [docs](https://prometheus.io/docs/prometheus/latest/configuration/configuration/#endpoints) 

```
If the endpoint is backed by a pod, all additional container ports of the pod, not bound to an endpoint port, are discovered as targets as well.
```

And  for prometheus this rule works
```
- job_name: foobar-internal-api
  scheme: http
  kubernetes_sd_configs:
    - role: endpoints
  relabel_configs:
  - source_labels: [__meta_kubernetes_namespace]
    action: keep
    regex: '^my_ns$'
  - source_labels: [__meta_kubernetes_endpoints_name]
    action: keep
    regex: '^foobar-internal-api$'
  - source_labels: [__meta_kubernetes_pod_container_port_number]
    action: keep
    regex: "5050"

- job_name: foobar-external-api
  scheme: http
  kubernetes_sd_configs:
    - role: endpoints
  relabel_configs:
  - source_labels: [__meta_kubernetes_namespace]
    action: keep
    regex: '^my_ns$'
  - source_labels: [__meta_kubernetes_endpoints_name]
    action: keep
    regex: '^foobar-external-api$'
  - source_labels: [__meta_kubernetes_pod_container_port_number]
    action: keep
    regex: "5050"
```
I've got my pods.

But for VMAgent, it contain all labels from endpoint, but only for port which bound in that endpoint
e.g.
```
{
__address__="10.15.181.52:32500", 
__meta_kubernetes_endpoint_address_target_kind="Pod", 
__meta_kubernetes_endpoint_address_target_name="foobar-external-api-65ffbb9cd9-xw4mc", __meta_kubernetes_endpoint_node_name="xxxxxxl", 
__meta_kubernetes_endpoint_port_name="external-api-service", 
__meta_kubernetes_endpoint_port_protocol="TCP", 
__meta_kubernetes_endpoint_ready="true", __meta_kubernetes_endpoints_annotation_endpoints_kubernetes_io_last_change_trigger_time="xxxxxx", __meta_kubernetes_endpoints_annotationpresent_endpoints_kubernetes_io_last_change_trigger_time="true", __meta_kubernetes_endpoints_label_app="foobar-external-api", __meta_kubernetes_endpoints_label_app_kubernetes_io_instance="foobar-gateway-eu", __meta_kubernetes_endpoints_labelpresent_app="true", __meta_kubernetes_endpoints_labelpresent_app_kubernetes_io_instance="true", __meta_kubernetes_endpoints_name="foobar-external-api", 
__meta_kubernetes_namespace="my_ns", __meta_kubernetes_pod_annotation_cni_projectcalico_org_podIP="xxx.xxx.xxx.xxx/32", __meta_kubernetes_pod_annotation_cni_projectcalico_org_podIPs="xxx.xxx.xxx.xxx/32", __meta_kubernetes_pod_annotationpresent_cni_projectcalico_org_podIP="true", __meta_kubernetes_pod_annotationpresent_cni_projectcalico_org_podIPs="true", __meta_kubernetes_pod_container_image="xxxxxxx, 
__meta_kubernetes_pod_container_name="foobar-external-api", 
__meta_kubernetes_pod_container_port_name="http", 
__meta_kubernetes_pod_container_port_number="32500", 
__meta_kubernetes_pod_container_port_protocol="TCP", 
__meta_kubernetes_pod_controller_kind="ReplicaSet", 
__meta_kubernetes_pod_controller_name="foobar-external-api-65ffbb9cd9", 
__meta_kubernetes_pod_host_ip="xxxxx", 
__meta_kubernetes_pod_ip="10.15.181.52", 
__meta_kubernetes_pod_label_app="foobar-external-api", __meta_kubernetes_pod_label_pod_template_hash="65ffbb9cd9", 
__meta_kubernetes_pod_labelpresent_app="true", 
__meta_kubernetes_pod_labelpresent_pod_template_hash="true", 
__meta_kubernetes_pod_name="foobar-external-api-65ffbb9cd9-xw4mc", 
__meta_kubernetes_pod_node_name="xxxxxxxx", 
__meta_kubernetes_pod_phase="Running", 
__meta_kubernetes_pod_ready="true", 
__meta_kubernetes_pod_uid="xxxxxxxxx", __meta_kubernetes_service_annotation_kubectl_kubernetes_io_last_applied_configuration="xxxx", __meta_kubernetes_service_annotationpresent_kubectl_kubernetes_io_last_applied_configuration="true", __meta_kubernetes_service_cluster_ip="xxxxxxx", 
__meta_kubernetes_service_label_app="foobar-external-api", __meta_kubernetes_service_label_app_kubernetes_io_instance="xxxxxxx", __meta_kubernetes_service_labelpresent_app="true", __meta_kubernetes_service_labelpresent_app_kubernetes_io_instance="true", __meta_kubernetes_service_name="foobar-external-api", 
__meta_kubernetes_service_type="NodePort", 
__metrics_path__="/metrics", 
__scheme__="http", 
__scrape_interval__="15s", 
__scrape_timeout__="10s", 
job="foobar-external-api"}	
```
This target has all labels but it's dropped because is wrong port, and it right.
And for other ports on pod:
```
{
__address__="10.15.181.52:5050", 
__meta_kubernetes_namespace="my_ns", __meta_kubernetes_pod_annotation_cni_projectcalico_org_podIP="xxx.xxx.xxx.xxx/32", __meta_kubernetes_pod_annotation_cni_projectcalico_org_podIPs="xxx.xxx.xxx.xxx/32", __meta_kubernetes_pod_annotationpresent_cni_projectcalico_org_podIP="true", __meta_kubernetes_pod_annotationpresent_cni_projectcalico_org_podIPs="true", __meta_kubernetes_pod_container_image="xxxxxxxxx", 
__meta_kubernetes_pod_container_name="foobar-external-api", 
__meta_kubernetes_pod_container_port_name="service", 
__meta_kubernetes_pod_container_port_number="5050", 
__meta_kubernetes_pod_container_port_protocol="TCP", 
__meta_kubernetes_pod_controller_kind="ReplicaSet", 
__meta_kubernetes_pod_controller_name="foobar-external-api-65ffbb9cd9", __meta_kubernetes_pod_host_ip="xxx.xxx.xxx.xxx", 
__meta_kubernetes_pod_ip="xxx.xxx.xxx.xxx", 
__meta_kubernetes_pod_label_app="foobar-external-api", __meta_kubernetes_pod_label_pod_template_hash="65ffbb9cd9", 
__meta_kubernetes_pod_labelpresent_app="true", 
__meta_kubernetes_pod_labelpresent_pod_template_hash="true", 
__meta_kubernetes_pod_name="foobar-external-api-65ffbb9cd9-xw4mc", __meta_kubernetes_pod_node_name="xxx.xxx.xxx.xxx", 
__meta_kubernetes_pod_phase="Running", 
__meta_kubernetes_pod_ready="true", 
__meta_kubernetes_pod_uid="f8314108-d529-47d4-b810-65efa832dec5", __meta_kubernetes_service_annotation_kubectl_kubernetes_io_last_applied_configuration="{\"apiVersion\":\"v1\"}", __meta_kubernetes_service_annotationpresent_kubectl_kubernetes_io_last_applied_configuration="true", __meta_kubernetes_service_cluster_ip="xxx.xxx.xxx.xxx", 
__meta_kubernetes_service_label_app="foobar-external-api", __meta_kubernetes_service_label_app_kubernetes_io_instance="xxxxx_xxxxx", __meta_kubernetes_service_labelpresent_app="true", __meta_kubernetes_service_labelpresent_app_kubernetes_io_instance="true", __meta_kubernetes_service_name="foobar-external-api", 
__meta_kubernetes_service_type="NodePort", 
__metrics_path__="/metrics", 
__scheme__="http", 
__scrape_interval__="15s", 
__scrape_timeout__="10s", 
job="foobar-external-api"}	
```
 this target dropped because of missing `__meta_kubernetes_endpoints_name`, and app endpoint labels as well
```
action: keep
source_labels: [__meta_kubernetes_endpoints_name]
regex: ^foobar-external-api$
```

And this one from Prometheus service discovery

```
__address__="10.15.181.52:5050"
__meta_kubernetes_endpoints_label_app="foobar-external-api"
__meta_kubernetes_endpoints_label_app_kubernetes_io_instance="xxxx"
__meta_kubernetes_endpoints_labelpresent_app="true"
__meta_kubernetes_endpoints_labelpresent_app_kubernetes_io_instance="true"
__meta_kubernetes_endpoints_name="foobar-external-api"
__meta_kubernetes_namespace="my_ns"
__meta_kubernetes_pod_annotation_cni_projectcalico_org_podIP="xxx.xxx.xxx.xxx/32"
__meta_kubernetes_pod_annotation_cni_projectcalico_org_podIPs="xxx.xxx.xxx.xxx/32"
__meta_kubernetes_pod_annotationpresent_cni_projectcalico_org_podIP="true"
__meta_kubernetes_pod_annotationpresent_cni_projectcalico_org_podIPs="true"
__meta_kubernetes_pod_container_name="foobar-external-api"
__meta_kubernetes_pod_container_port_name="service"
__meta_kubernetes_pod_container_port_number="5050"
__meta_kubernetes_pod_container_port_protocol="TCP"
__meta_kubernetes_pod_controller_kind="ReplicaSet"
__meta_kubernetes_pod_controller_name="foobar-external-api-65ffbb9cd9"
__meta_kubernetes_pod_host_ip="xxxxxx"
__meta_kubernetes_pod_ip="10.15.181.52"
__meta_kubernetes_pod_label_app="foobar-external-api"
__meta_kubernetes_pod_label_pod_template_hash="65ffbb9cd9"
__meta_kubernetes_pod_labelpresent_app="true"
__meta_kubernetes_pod_labelpresent_pod_template_hash="true"
__meta_kubernetes_pod_name="foobar-external-api-65ffbb9cd9-xw4mc"
__meta_kubernetes_pod_node_name="xxxx"
__meta_kubernetes_pod_phase="Running"
__meta_kubernetes_pod_ready="true"
__meta_kubernetes_pod_uid="xxxxx"
__meta_kubernetes_service_annotation_kubectl_kubernetes_io_last_applied_configuration="xxxxxxxx"
__meta_kubernetes_service_annotationpresent_kubectl_kubernetes_io_last_applied_configuration="true"
__meta_kubernetes_service_label_app="foobar-external-api"
__meta_kubernetes_service_label_app_kubernetes_io_instance="foobar-gateway-eu"
__meta_kubernetes_service_labelpresent_app="true"
__meta_kubernetes_service_labelpresent_app_kubernetes_io_instance="true"
__meta_kubernetes_service_name="foobar-external-api"
__metrics_path__="/metrics"
__scheme__="http"
__scrape_interval__="15s"
__scrape_timeout__="10s"
job="foobar-external-api"
```
Here as you can see all labels are present on pod with port which not bound to endpoint

**Comment by zekker6** [comment]
@thunderbird86 Thank you for the details!

Do I understand right that your configuration is similar to the following:
```
apiVersion: apps/v1
kind: Deployment
metadata:
  name: nginx-multiport
  labels:
    app: nginx-multiport
spec:
  replicas: 2
  template:
    metadata:
      labels:
        app: nginx-multiport
    spec:
      containers:
        - name: nginx
          image: nginx:alpine
          ports:
            - containerPort: 80
              name: http
            - containerPort: 443
  selector:
    matchLabels:
      app: nginx-multiport
---
apiVersion: v1
kind: Service
metadata:
  name: nginx-multiport
spec:
    selector:
        app: nginx-multiport
    ports:
      - name: http
        port: 80
        targetPort: 80
```
So that pod has 2 ports, but service only uses one of those?

As far as I can see this is the only option to get it discovered, but missing endpoint name because it does not belong to any endpoint.
Prometheus uses additional logic to add metadata for such pods:
https://github.com/prometheus/prometheus/blob/0ab95536115adfe50af249d36d73674be694ca3f/discovery/kubernetes/endpoints.go#L264-L404

We will implement similar logic on the VMAgent side to make it consistent with Prometheus.

**Comment by zekker6** [comment]
@thunderbird86  Here is an image which is built from #4235 : `docker.io/victoriametrics/vmagent:heads-fix-vmagent-add-epp-name-0-gd548882ce`.
It would be great if you could test that it fixes discovery inconsistency for you.

**Comment by thunderbird86** [comment]
Hello @zekker6.

Thanks, it works. But it also missing some labels, which could be issue in future.

Here is target from Prometheus 
```
{
  "discoveredLabels": {
    "__address__": "10.15.181.53:8080",
    "__meta_kubernetes_endpoints_label_app": "foobar-internal-api",
    "__meta_kubernetes_endpoints_label_app_kubernetes_io_instance": "xxxxxx",
    "__meta_kubernetes_endpoints_labelpresent_app": "true",
    "__meta_kubernetes_endpoints_labelpresent_app_kubernetes_io_instance": "true",
    "__meta_kubernetes_endpoints_name": "foobar-internal-api",
    "__meta_kubernetes_namespace": "my_ns",
    "__meta_kubernetes_pod_annotation_cni_projectcalico_org_podIP": "x",
    "__meta_kubernetes_pod_annotation_cni_projectcalico_org_podIPs": "x",
    "__meta_kubernetes_pod_annotationpresent_cni_projectcalico_org_podIP": "true",
    "__meta_kubernetes_pod_annotationpresent_cni_projectcalico_org_podIPs": "true",
    "__meta_kubernetes_pod_container_name": "foobar-internal-api",
    "__meta_kubernetes_pod_container_port_name": "service",
    "__meta_kubernetes_pod_container_port_number": "8080",
    "__meta_kubernetes_pod_container_port_protocol": "TCP",
    "__meta_kubernetes_pod_controller_kind": "ReplicaSet",
    "__meta_kubernetes_pod_controller_name": "foobar-internal-api-6f8d6fc9dc",
    "__meta_kubernetes_pod_host_ip": "x",
    "__meta_kubernetes_pod_ip": "10.15.181.53",
    "__meta_kubernetes_pod_label_app": "foobar-internal-api",
    "__meta_kubernetes_pod_label_pod_template_hash": "6f8d6fc9dc",
    "__meta_kubernetes_pod_labelpresent_app": "true",
    "__meta_kubernetes_pod_labelpresent_pod_template_hash": "true",
    "__meta_kubernetes_pod_name": "foobar-internal-api-6f8d6fc9dc-qvxkp",
    "__meta_kubernetes_pod_node_name": "x",
    "__meta_kubernetes_pod_phase": "Running",
    "__meta_kubernetes_pod_ready": "true",
    "__meta_kubernetes_pod_uid": "x",
    "__meta_kubernetes_service_annotation_kubectl_kubernetes_io_last_applied_configuration": "x",
    "__meta_kubernetes_service_annotationpresent_kubectl_kubernetes_io_last_applied_configuration": "true",
    "__meta_kubernetes_service_label_app": "foobar-internal-api",
    "__meta_kubernetes_service_label_app_kubernetes_io_instance": "foobar-gateway-eu",
    "__meta_kubernetes_service_labelpresent_app": "true",
    "__meta_kubernetes_service_labelpresent_app_kubernetes_io_instance": "true",
    "__meta_kubernetes_service_name": "foobar-internal-api",
    "__metrics_path__": "/metrics",
    "__scheme__": "http",
    "__scrape_interval__": "15s",
    "__scrape_timeout__": "10s",
    "job": "foobar-internal-api"
  },
  "labels": {
    "instance": "100.115.181.53:8080",
    "job": "foobar-internal-api"
  },
  "scrapePool": "marketo-internal-api",
  "scrapeUrl": "http://10.15.181.53:8080/metrics",
  "globalUrl": "http://10.15.181.53:8080/metrics",
  "lastError": "",
  "lastScrape": "x",
  "lastScrapeDuration": 0.002710549,
  "health": "up",
  "scrapeInterval": "15s",
  "scrapeTimeout": "10s"
}
```

As far as you can see there, also persent following labels 

`__meta_kubernetes_endpoints_label_<labelname>`
`__meta_kubernetes_endpoints_labelpresent_<labelname>`

Which is missing with `docker.io/victoriametrics/vmagent:heads-fix-vmagent-add-epp-name-0-gd548882ce`

I've fixed this using following code. Not sure how to create [PR](https://github.com/VictoriaMetrics/VictoriaMetrics/pull/4245), because I use LTS version.

**Comment by zekker6** [comment]
@thunderbird86 Thanks for the PR, changes looks good to me. Just want to ask if you could [sign](https://docs.github.com/en/authentication/managing-commit-signature-verification/signing-commits) your commit?
You'll need to make a PR into master branch, we're backporting bug fixes into LTS branches as a part of release process so you don't need to do something special for this to happen.

**Comment by valyala** [comment]
FYI, the bugfix has been included in [vmagent v1.91.0](https://docs.victoriametrics.com/CHANGELOG.html)

**Comment by valyala** [comment]
The bugfix has been backported to [v1.87.6 LTS release](https://github.com/VictoriaMetrics/VictoriaMetrics/releases/tag/v1.87.6).
